### PR TITLE
stack error (underflow & overflow) circuit

### LINKED ISSUE #910: implement stack error(overflow & underflow)  circuit
PR #873 targets this issue .

### PR DESCRIPTION
**Pull request by DreamWuGit**
spec PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/301
  - [x] merge stack overflow and underflow error to one `ErrorStack`
  - [x] construct op code stack table and circuit constraint
  - [x]  overflow and underflow tests in root/internal calls

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203326967390172

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    #[test]
+    fn pop_gadget_underflow() {
+        test_stack_underflow(Word::from(0x030201));
+        test_stack_underflow(Word::from(0xab));
+    }
+
+    #[test]
+    fn stack_overflow_simple() {
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/execution/error_stack.rs:196]: `stack_overflow_simple` seems redundant as `test_stack_overflow` is called only at one place.

**Review by miha-stopar — APPROVED** [review]
LGTM! There are just a couple of nitpicks that might be resolved.

File: eth-types/src/evm_types/opcode_ids.rs [hunk 2]
```diff
@@ -663,6 +663,159 @@ impl OpcodeId {
         }
     }
 
+    /// Returns the constant min & stack pointer of `OpcodeId`
+    pub const fn stack_pair(&self) -> (u32, u32) {
+        match self {
+            // min_stack 0 means stack overflow never happen, for example, `OpcodeId::ADD` only
+            // possilbe encounter underflow error, but never encounter overflow error.
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply eth-types/src/evm_types/opcode_ids.rs:670]: will add explanation for `max_stack`

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_stack.rs:196]: thanks, will check it .

File: eth-types/src/evm_types/opcode_ids.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply eth-types/src/evm_types/opcode_ids.rs]: updated in [fea873d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/fea873dcaa371498fccfda47be88760437797378) thanks!

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_stack.rs:196]: added more two calls in `stack_overflow_simple` for different push, updated in `stack_overflow_simple` [fea873d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/fea873dcaa371498fccfda47be88760437797378) thanks!

**Comment by ChihChengLiang** [comment]
@icemelon can you assign a reviewer for this one?

**Comment by icemelon** [comment]
> @icemelon can you assign a reviewer for this one?

@lispc could you help review this PR?

File: eth-types/src/evm_types/opcode_ids.rs [hunk 3]
```diff
@@ -663,6 +663,162 @@ impl OpcodeId {
         }
     }
 
+    /// Returns the constant min & stack pointer of `OpcodeId`
+    pub const fn stack_pair(&self) -> (u32, u32) {
```
Comment (lispc) [inline eth-types/src/evm_types/opcode_ids.rs:667]: what about rename this to something like "valid_stack_ptr_range_before_exec"?   self-explainable.

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+            min_stack_pointer.expr(),
+        );
+        let is_underflow = LtGadget::construct(
+            cb,
+            max_stack_pointer.expr(),
+            cb.curr.state.stack_pointer.expr(),
+        );
+
+        cb.require_boolean("is_overflow is bool", is_overflow.expr());
+        cb.require_boolean("is_underflow is bool", is_underflow.expr());
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/error_stack.rs:62]: don't need require_boolean here? LtGadget already called cb.query_bool inside.

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+        // current call must be failed.
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 0.expr());
+
+        // Go to EndTx only when is_root
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr(),
+            is_to_end_tx,
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/error_stack.rs:80]: `require_equal` is used. So "Go to EndTx if and only if is_root"?

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        let min_stack = opcode.stack_pair().0 as u64;
+        let max_stack = opcode.stack_pair().1 as u64;
+
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/error_stack.rs:130]: more rust style?

let (min_stack, max_stack) = opcode.stack_pair();
self.min_stack_pointer
            .assign(region, offset, Value::known(F::from(min_stack as u64)))?;
self.max_stack_pointer
    .assign(region, offset, Value::known(F::from(max_stack as u64)))?;

File: zkevm-circuits/src/evm_circuit/step.rs [hunk 7]
```diff
@@ -86,8 +86,8 @@ pub enum ExecutionState {
     SELFDESTRUCT,
     // Error cases
     ErrorInvalidOpcode,
-    ErrorStackOverflow,
-    ErrorStackUnderflow,
+    ErrorStack,
+    //ErrorStackUnderflow,
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/step.rs:90]: //ErrorStackUnderflow, can be removed

File: zkevm-circuits/src/evm_circuit/param.rs [hunk 8]
```diff
@@ -56,5 +56,7 @@ pub(crate) const N_BYTES_PROGRAM_COUNTER: usize = N_BYTES_U64;
 // Number of bytes that will be used for a tx's gas field.
 pub(crate) const N_BYTES_GAS: usize = N_BYTES_U64;
 
+pub(crate) const N_BYTES_STACK: usize = 2;
+
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/param.rs:60]: i think this is only useful in ErrorStackGadget. So maybe put it in that file is good enough. Don't need to be put it  high level.

File: eth-types/src/evm_types/opcode_ids.rs (hunk 3 above)
Comment (DreamWuGit) [reply eth-types/src/evm_types/opcode_ids.rs:667]: `valid_stack_ptr_range` good name !

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs (hunk 6 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_stack.rs:130]: good suggestion, I like it !

File: eth-types/src/evm_types/opcode_ids.rs (hunk 3 above)
Comment (DreamWuGit) [reply eth-types/src/evm_types/opcode_ids.rs]: updated in [3c9c55d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/3c9c55da2df49e8f6a236241b8d42fcfe6f4f465) thanks!

File: zkevm-circuits/src/evm_circuit/execution/error_stack.rs (hunk 6 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_stack.rs:129]: updated in [3c9c55d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/3c9c55da2df49e8f6a236241b8d42fcfe6f4f465) thanks!

File: zkevm-circuits/src/evm_circuit/param.rs (hunk 8 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/param.rs]: move to gadget level in [3c9c55d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/3c9c55da2df49e8f6a236241b8d42fcfe6f4f465) thanks!

File: zkevm-circuits/src/evm_circuit/step.rs (hunk 7 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/step.rs]: removed in [3c9c55d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/3c9c55da2df49e8f6a236241b8d42fcfe6f4f465) thanks!

**Review by lispc — APPROVED** [review]

**Pull request merged**